---a/net/minecraft/entity/projectile/EntitySmallFireball.java
+++b/net/minecraft/entity/projectile/EntitySmallFireball.java
@@ -8,63 +8,62 @@
 import net.minecraft.util.MovingObjectPosition;
 import net.minecraft.world.World;
 
-public class EntitySmallFireball extends EntityFireball
-{
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+
+public class EntitySmallFireball extends EntityFireball {
+
     private static final String __OBFID = "CL_00001721";
 
-    public EntitySmallFireball(World worldIn)
-    {
+    public EntitySmallFireball(World worldIn) {
         super(worldIn);
         this.setSize(0.3125F, 0.3125F);
     }
 
-    public EntitySmallFireball(World worldIn, EntityLivingBase p_i1771_2_, double p_i1771_3_, double p_i1771_5_, double p_i1771_7_)
-    {
+    public EntitySmallFireball(World worldIn, EntityLivingBase p_i1771_2_, double p_i1771_3_, double p_i1771_5_, double p_i1771_7_) {
         super(worldIn, p_i1771_2_, p_i1771_3_, p_i1771_5_, p_i1771_7_);
         this.setSize(0.3125F, 0.3125F);
     }
 
-    public EntitySmallFireball(World worldIn, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_)
-    {
+    public EntitySmallFireball(World worldIn, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_) {
         super(worldIn, p_i1772_2_, p_i1772_4_, p_i1772_6_, p_i1772_8_, p_i1772_10_, p_i1772_12_);
         this.setSize(0.3125F, 0.3125F);
     }
 
-    protected void onImpact(MovingObjectPosition movingObject)
-    {
-        if (!this.worldObj.isRemote)
-        {
+    protected void onImpact(MovingObjectPosition movingObject) {
+        if (!this.worldObj.isRemote) {
             boolean flag;
 
-            if (movingObject.entityHit != null)
-            {
+            if (movingObject.entityHit != null) {
                 flag = movingObject.entityHit.attackEntityFrom(DamageSource.causeFireballDamage(this, this.shootingEntity), 5.0F);
 
-                if (flag)
-                {
+                if (flag) {
                     this.func_174815_a(this.shootingEntity, movingObject.entityHit);
 
-                    if (!movingObject.entityHit.isImmuneToFire())
-                    {
-                        movingObject.entityHit.setFire(5);
+                    if (!movingObject.entityHit.isImmuneToFire()) {
+                        // CraftBukkit start - Entity damage by entity event + combust event
+                        EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), movingObject.entityHit.getBukkitEntity(), 5);
+                        movingObject.entityHit.worldObj.getServer().getPluginManager().callEvent(event);
+
+                        if (!event.isCancelled()) {
+                            movingObject.entityHit.setFire(event.getDuration());
+                        }
+                        // CraftBukkit end
                     }
                 }
-            }
-            else
-            {
+            } else {
                 flag = true;
 
-                if (this.shootingEntity != null && this.shootingEntity instanceof EntityLiving)
-                {
+                if (this.shootingEntity != null && this.shootingEntity instanceof EntityLiving) {
                     flag = this.worldObj.getGameRules().getGameRuleBooleanValue("mobGriefing");
                 }
 
-                if (flag)
-                {
+                if (flag) {
                     BlockPos blockpos = movingObject.getBlockPos().offset(movingObject.sideHit);
 
-                    if (this.worldObj.isAirBlock(blockpos))
-                    {
-                        this.worldObj.setBlockState(blockpos, Blocks.fire.getDefaultState());
+                    if (this.worldObj.isAirBlock(blockpos)) {
+                        // CraftBukkit start
+                        if (isIncendiary && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(worldObj, blockpos.getX(), blockpos.getY(), blockpos.getZ(), this).isCancelled()) {
+                            this.worldObj.setBlockState(blockpos, Blocks.fire.getDefaultState());
+                        }
+                        // CraftBukkit end
                     }
                 }
             }
@@ -74,14 +74,11 @@
         }
     }
 
-    public boolean canBeCollidedWith()
-    {
+    public boolean canBeCollidedWith() {
         return false;
     }
 
-    public boolean attackEntityFrom(DamageSource source, float amount)
-    {
+    public boolean attackEntityFrom(DamageSource source, float amount) {
         return false;
     }
-}
+}
